File: src/cmd_parser.c
#include "../include/cmd_parser.h"
#include "../include/makefile_gen.h"
#include "../include/dir_setup.h"

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <getopt.h>
#include <stdarg.h>

#define CMD_LEN (sizeof(commands)/sizeof(commands[0]))
command_t commands[] = {
    {"help", cmd_help},
	{"init", cmd_init},
    {"set-default", cmd_set_default},
};

static struct option flag_options[] = {
	{"lang", required_argument, 0, 'l'},
	{"format", required_argument, 0, 'f'},
	{"dirs", required_argument, 0, 'd'},
};

int cmd_help(int argc, char** argv) {
    puts("Usage: rm [CMD] [OPTION]... [DIR]");
    puts("Setup Command-LIne Development Environment in DIR.");

    printf("  \033[1;94m-l,  --lang\033[0m,\n");
    puts("        specify the language of the target DE");

    printf("  \033[1;94m-f,  --format\033[0m,\n");
    puts("        specify the target output file format");

    printf("  \033[1;94m-d,  --dirs\033[0m,\n");
    puts("        specify the directory layout of the target DE");


    return 0;
}

int cmd_init(int argc, char** argv) {
	if (argc < 3 || argv[2][0] == '-') {
		fprintf(stderr, "Init: missing source directory\nTry \'cenv init --help\' for more information.\n");
		exit(1);
	}	

	char path[256];
   
    opt_t opts;

    int opt = 0;

    while ((opt = getopt_long(argc, argv, "l:f:d:h", flag_options, NULL)) != -1) {
		switch (opt) {
			case 'l':
			    opts.lang = optarg;
				break;
			case 'f':
				opts.format = optarg;
				break;
			case 'd':
				opts.dirs =  optarg;
				break;
			default:
				fprintf(stderr, "Init: invalid option \'%c\'\nTry \'cenv init --help\' for more information.\n", opt);
				exit(1);	
		}
	} 

	if (!opts.lang || !opts.format || !opts.dirs) {
		fprintf(stderr, "Init: must specify language, format, and directory layout\nTry \'cenv init --help\' for more information.\n");
		exit(1);
	}

	setup_dir(opts.dirs);
	snprintf(path, sizeof(path), "templates/makefiles/%s/%s/%s", opts.lang, opts.format, opts.dirs);
	load_template(path);
	
	return 0;
}

int cmd_set_default (int argc, char** argv) {


}

int parse_commands(int argc, char** argv) {
	for (size_t i = 0; i < CMD_LEN; i++) {
		if (strcmp(argv[1], commands[i].name) == 0) {
			return commands[i].cmd_handler(argc, argv);	
		}	
	}	

	return 0;
}

File: rust/src/cmd_parser.rs
use crate::dir_setup::setup_dir;
use crate::makefile_gen::load_makefile;

use phf::phf_map;
use std::process::exit;
extern crate getopts;
use getopts::Options;

#[derive(Debug)]
pub enum CmdError {
    MissingFlag(String),
    ParseError(String),
}

impl std::fmt::Display for CmdError {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        match self {
            CmdError::MissingFlag(flag) => write!(f, "Error, missing flag --{}", flag),
            CmdError::ParseError(msg) => write!(f, "Parse Error:  {}", msg),
        }
    }
}

fn require_opt(matches: &getopts::Matches, flag: &str) -> Result<String, CmdError> {
    matches 
        .opt_str(flag)
        .ok_or_else(|| CmdError::MissingFlag(flag.to_string()))
}

fn cmd_help(_args: Vec<String>) -> Result<i8, CmdError> {
    println!("Usage: rm [CMD] [OPTION]... [DIR]");
    println!("Setup Command-LIne Development Environment in DIR.");

    println!("  \x1b[1m-l,  --lang\x1b[0m");
    println!("         specify the language of the target DE\n");

    println!("  \x1b[1m-f,  --format\x1b[0m");
    println!("         specify the target output file format\n");

    println!("  \x1b[1m-d,  --dirs\x1b[0m");
    println!("         specify the directory layout of the target DE\n"); 

    Ok(0)
}

fn cmd_init(args: Vec<String>) -> Result<i8, CmdError>  {  
    let mut opts = Options::new();  
    opts.optopt("l", "lang", "specify the language of the target DE", "LANGUAGE");
    opts.optopt("f", "format", "specify the target output file format", "FORMAT");
    opts.optopt("d", "dirs", "specify the directory layout of the target DE", "DIR_LAYOUT");
    
    let matches = opts
        .parse(&args[2..])
        .map_err(|e| CmdError::ParseError(e.to_string()))?;

    let required: &[&str] = &["l", "f", "d"];
    for flag in required {
        require_opt(&matches, flag)?;
    }

    let lang = require_opt(&matches, "l")?;
    let format = require_opt(&matches, "f")?;
    let dirs = require_opt(&matches, "d")?;

    let path = format!("makefiles/{lang}/{format}/{dirs}");
    println!("{}", path);  

    setup_dir(dirs, args[2].clone());
    load_makefile(path, args[2].clone());

    Ok(0)
}

fn cmd_default(args: Vec<String>) -> Result<i8, CmdError> {
    if args[2] == "".to_string() {
        
    }
    Ok(0)
}

fn cmd_set_default(args: Vec<String>) -> Result<i8, CmdError> {
    if args[2] == "".to_string() {
    
    }
    Ok(0)
}

static COMMANDS: phf::Map<&'static str, fn(Vec<String>) -> Result<i8, CmdError>> = phf_map! {
    "help" => cmd_help as fn(Vec<String>) -> Result<i8, CmdError>,
    "init" => cmd_init as fn(Vec<String>) -> Result<i8, CmdError>,
    "default" => cmd_default as fn(Vec<String>) -> Result<i8, CmdError>,
    "set_default" => cmd_set_default as fn(Vec<String>) -> Result<i8, CmdError>,
};

pub fn parse_commands(args: Vec<String>) -> i8 {
    if let Some(cmd_helper) = COMMANDS.get(&args[1]) {
        match cmd_helper(args) {
            Ok(_) => {}
            Err(e) => {
                eprintln!("{}", e);
                std::process::exit(1);
            }
        }
    } else {
        // Make better
        println!("Error: invalid command \'{}\'", args[1]);
        exit(1);
    }   

    return 0;
}


